python/cudf/cudf/core/dataframe.py

```diff
@@ -1453,84 +1420,83 @@ def _get_columns_by_label(self, labels, downcast=False):
         )
         return out
 
-    # unary, binary, rbinary, orderedcompare, unorderedcompare
-    def _apply_op(self, fn, other=None, fill_value=None):
-
-        result = DataFrame(index=self.index)
-
-        def op(lhs, rhs):
-            if fill_value is None:
-                return getattr(lhs, fn)(rhs)
-            else:
-                return getattr(lhs, fn)(rhs, fill_value)
-
-        if other is None:
-            for col in self._data:
-                result[col] = getattr(self[col], fn)()
-        elif isinstance(other, Sequence):
-            # This adds the ith element of other to the ith column of self.
-            for k, col in enumerate(self._data):
-                result[col] = getattr(self[col], fn)(other[k])
-        elif isinstance(other, DataFrame):
+    def _binaryop(
+        self, other, fn, fill_value=None, reflect=False,
+    ):
+        lhs, rhs = self, other
+
+        if _is_scalar_or_zero_d_array(rhs):
+            rhs = [rhs] * lhs._num_columns
+
+        # For columns that exist in rhs but not lhs, we swap the order so that
+        # we can always assume that left has a binary operator. This
```

Comment:
Could you help me understand this with an example?

Reply:
Yes, the relevant cases are like what you see [here](https://github.com/rapidsai/cudf/pull/8598/files#diff-72a734c5cdcce0fc7c089af79b44da4aa2a3f325a34fbfb0288565938b03e252R1465-R1471). Consider something like
`cudf.DataFrame({'a': [1, 2, 3]}).add(cudf.DataFrame({'b': [4, 5, 6]}), fill_value=42)`. The result needs to contain a column `output['b'] = [46, 47, 48]`. The assumption made by `Frame._colwise_binop` is that we can always invoke `lhs.binary_operator(...)` as is done [here](https://github.com/rapidsai/cudf/pull/8598/files#diff-45f163ddc7fb3fc5853e1fdd90a3d740a9dd0987c9ebcf00315ad7aff727576dR3451-R3459). In the case where there is no rhs the new code bypasses the binop altogether and just generates a null column, but if there's a fill value we need to do a binop between the column and a scalar (reflected as needed).

Reply:
FWIW I'm all for simplifying this implementation even further if you can think of good ways to do so. The number of different edge cases to be considered and use cases to support are the main bottlenecks to further simplification.